vanilla/src/main/java/org/spongepowered/vanilla/world/VanillaWorldManager.java

```diff
@@ -163,64 +157,295 @@ public ResourceKey getDefaultPropertiesKey() {
+        final Integer dimensionId = ((WorldInfoBridge) worldInfo).bridge$getDimensionId();
+
+        if (dimensionId == null) {
+            SpongeCommon.getLogger().error("World '{}' has no dimension id in the Sponge level data ({}). Aborting...", key, Constants.Sponge
+                    .World.LEVEL_SPONGE_DAT);
+            return CompletableFuture.completedFuture(Optional.empty());
+        }
+
+        final ResourceKey existingKey = ((ResourceKeyBridge) worldInfo).bridge$getKey();
+        if (existingKey != null && !existingKey.equals(key)) {
+            SpongeCommon.getLogger().error("World '{}' is keyed as '{}' in the level data. Aborting...", key, existingKey);
+            return CompletableFuture.completedFuture(Optional.empty());
+        }
+
+        ((ResourceKeyBridge) worldInfo).bridge$setKey(key);
+
+        final SpongeDimensionType logicType = ((WorldInfoBridge) worldInfo).bridge$getLogicType();
+
+        final DimensionType dimensionType = Registry.DIMENSION_TYPE.getValue((ResourceLocation) (Object) key).orElseGet(() -> this.
+                createDimensionType(key, logicType, worldDirectory.getFileName().toString(), dimensionId + 1));
+
+        if (dimensionType.getId() != dimensionId) {
+            SpongeCommon.getLogger().error("World '{}' specifies internal id '{}' which was already registered. Aborting...", key, dimensionId);
+            return CompletableFuture.completedFuture(Optional.empty());
+        }
+
+        ((DimensionTypeBridge) dimensionType).bridge$setSpongeDimensionType(logicType);
+
+        MinecraftServerAccessor_Vanilla.accessor$getLogger().info("Loading World '{}' ({}/{})", key, logicType.getKey().getFormatted(), dimensionType.getId());
+
+        final IChunkStatusListener chunkStatusListener = ((MinecraftServerAccessor_Vanilla) this.server).accessor$getChunkStatusListenerFactory().create(11);
+
+        world = new ServerWorld(this.server, this.server.getBackgroundExecutor(), saveHandler, worldInfo,
+                dimensionType, this.server.getProfiler(), chunkStatusListener);
+
+        this.loadedWorldInfos.put(key, worldInfo);
+        this.infoByType.put(dimensionType, worldInfo);
+        this.allInfos.put(key, worldInfo);
+        this.worlds.put(key, world);
+        this.worldsByType.put(dimensionType, world);
+
+        this.performPostLoadWorldLogic(world, this.createDefaultSettings(null, false, worldInfo.getSeed(), worldInfo.getGenerator(), null), worldDirectory, chunkStatusListener);
+
+        return CompletableFuture.completedFuture(Optional.of((org.spongepowered.api.world.server.ServerWorld) world));
     }
 
     @Override
-    public CompletableFuture<Optional<org.spongepowered.api.world.server.ServerWorld>> loadWorld(WorldProperties properties) throws IOException {
-        return null;
+    public CompletableFuture<Optional<org.spongepowered.api.world.server.ServerWorld>> loadWorld(WorldProperties properties) {
```

Comment:
I think someone else said this previously, not sure if it was about this specifically or somewhere else, but in all cases in this method an empty Optional is actually an exceptional condition. So, we could make this return `CompletableFuture<ServerWorld>` and call [`CompletableFuture#completeExceptionally`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#completeExceptionally-java.lang.Throwable-), which would then be caught by plugin developers via the checked exception from `get()` (or if they use `join`, then that's their lookout).

The downside of this is creating the exception (could be expensive due to stack trace generation), and knowing what exception we should be throwing - at least that way a plugin can get more context as to why a world did not load.